fendermint/vm/topdown/src/sync/syncer.rs

```diff
@@ -0,0 +1,268 @@
+        let height = self.sync_pointers.head() + 1;
+        let parent_block_hash = self.non_null_parent_hash().await;
+
+        let block_hash_res = match self.parent_proxy.get_block_hash(height).await {
+            Ok(res) => res,
+            Err(e) => {
+                let err = e.to_string();
+                if is_null_round_str(&err) {
+                    tracing::warn!(height, "null round at height");
+
+                    self.sync_pointers.advance_head();
+
+                    return Ok(None);
+                }
+                return Err(Error::CannotQueryParent(err, height));
+            }
+        };
+
+        if block_hash_res.parent_block_hash != parent_block_hash {
+            tracing::warn!(
+                height,
+                parent_hash = hex::encode(&block_hash_res.parent_block_hash),
+                previous_hash = hex::encode(&parent_block_hash),
+                "parent block hash diff than previous hash",
+            );
+            return Err(Error::ParentChainReorgDetected);
+        }
+
+        if let Some(to_confirm) = self.sync_pointers.to_confirm() {
+            tracing::warn!(
+                height,
+                confirm = to_confirm,
+                "non-null round at height, confirmed previous height"
+            );
+            let data = self
+                .fetch_data(to_confirm, block_hash_res.block_hash)
+                .await?;
+            self.sync_pointers.advance_confirm(height);
+            return Ok(Some((to_confirm, data)));
+        }
+
+        tracing::warn!(height, "non-null round at height, waiting for confirmation");
+        self.sync_pointers.advance_head();
+
+        Ok(None)
+    }
+
+    async fn fetch_data(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+    ) -> Result<ParentViewPayload, Error> {
+        let changes_res = self
+            .parent_proxy
+            .get_validator_changes(height)
+            .await
+            .map_err(|e| Error::CannotQueryParent(e.to_string(), height))?;
+        if changes_res.block_hash != block_hash {
+            tracing::warn!(
+                height,
+                change_set_hash = hex::encode(&changes_res.block_hash),
+                block_hash = hex::encode(&block_hash),
+                "change set block hash does not equal block hash",
+            );
+            return Err(Error::ParentChainReorgDetected);
+        }
+
+        let top_down_msgs_res = self
+            .parent_proxy
+            .get_top_down_msgs_with_hash(height, &block_hash)
+            .await
+            .map_err(|e| Error::CannotQueryParent(e.to_string(), height))?;
+
+        Ok((block_hash, changes_res.value, top_down_msgs_res))
+    }
+
+    /// We only want the non-null parent block's hash
+    async fn non_null_parent_hash(&self) -> BlockHash {
+        let parent_height = match (self.sync_pointers.to_confirm(), self.sync_pointers.tail()) {
```

Comment:
What about calling it `.pending_exec` instead of `.to_confirm`? The latter confuse=d me for a moment.
```suggestion
        let parent_height = match (self.sync_pointers.pending_exec(), self.sync_pointers.tail()) {
```